feat(Chip): Add the Chip component to PF4

boaz0 opened the pull request:
What: [#976](https://github.com/patternfly/patternfly-react/issues/976)

patternfly-build commented:
PatternFly-React preview: https://1076-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://1076-pr-patternfly-react-patternfly.surge.sh

karelhala commented on packages/patternfly-4/react-core/src/components/Chip/Chip.js:
```diff
@@ -0,0 +1,45 @@
+  return !isOverflowed ? (
+    <ChipWrapper className={className} {...props}>
+      <ChipText id={id}>{children}</ChipText>
+      <ChipButton aria-label="Remove" id={`remove_${id}`} aria-labelledby={`remove_${id} ${id}`}>
```
You are missing `onClick` for remove button. Clicking on chip anywhere should not call the on click function, this function should be called only for `ChipButton`.

boaz1337 replied:
Thanks! I can't believe I missed it. I will fix it soon as possible.

ibolton336 commented on packages/patternfly-4/react-core/src/components/Chip/examples/SimpleChip.js:
```diff
@@ -0,0 +1,41 @@
+  render() {
+    return (
+      <React.Fragment>
+        {this.state.show.includes('chip_one') && (
```
The component looks great - nice work. Just a few things - can we leverage destructuring here for cleaning up the render method of state references?

boaz1337 replied:
Yeah sure!

ibolton336 commented on packages/patternfly-4/react-core/src/components/Chip/ChipWrapper.js:
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
```
Can we use typescript for this component?

boaz1337 replied:
I thought `ChipWrapper.d.ts` was good enough. Can you give me an example of doing that? Like rename the file from `ChipWrapper.js` to `ChipWrapper.ts` or something more complicated?

Thanks :smile_cat:

ibolton336 replied:
The typescript stuff isn't fully baked yet for PF4 so I think your are right to just include the d.ts file. Eventually we will want more but for now this should be fine.

boaz1337 replied:
Cool :sunglasses: . At any rate, if I can help with that at any point let me know.

ibolton336 commented on packages/patternfly-4/react-core/src/components/Chip/Chip.js:
```diff
@@ -0,0 +1,45 @@
+import { TimesCircleIcon } from '@patternfly/react-icons';
+
+const Chip = ({ isOverflowed, children, id, className, ...props }) => {
+  return !isOverflowed ? (
```
Can we use a different type of conditional rendering here that would lend itself to more than 2 types of chips? That would help with future proofing the component.

boaz1337 replied:
I am not sure I understand what you mean.

I can do this differently in the following ways:

```js
const chipBody = !isOverflowed ? (
 <ChipText id={id}>{children}</ChipText>
      <ChipButton aria-label="Remove" id={`remove_${id}`} aria-labelledby={`remove_${id} ${id}`}>
        <TimesCircleIcon aria-hidden="true" />
      </ChipButton>
  ) : (
      <ChipButton aria-label={children}>
        <ChipText>{children}</ChipText>
      </ChipButton>
);

return <ChipWrapper isOverflowed={isOverflowed} className={className} {...props}>{chipBody}</ChipWrapper>
```
or maybe:

```js
const { chipType } = this.props;
switch(chipType) {
case "overflow":
  return (
    <ChipWrapper isOverflowed className={className} {...props}>
      <ChipButton aria-label={children}>
        <ChipText>{children}</ChipText>
      </ChipButton>
    </ChipWrapper>
  );
default:
  return (
    <ChipWrapper className={className} {...props}>
      <ChipText id={id}>{children}</ChipText>
      <ChipButton aria-label="Remove" id={`remove_${id}`} aria-labelledby={`remove_${id} ${id}`}>
        <TimesCircleIcon aria-hidden="true" />
      </ChipButton>
    </ChipWrapper>
  );
}
```

Let me know if you have a better or a different idea. Thanks :fish_cake:

ibolton336 replied:
I'd take a look at the current patterns used within the library. Maybe something like the button component "variant" prop approach?

tlabaj commented on packages/patternfly-4/react-core/src/components/Chip/Chip.js:
```diff
@@ -0,0 +1,57 @@
+  /** Content rendered inside the chip text */
+  children: PropTypes.string.isRequired,
+  /** ID of the chip */
+  id: props => {
```
In other components, if the consumer does not provide an id that is required for accessibility,  we  have been generating it internally.  Maybe we should consider doing that here and getting rid of this check.

tlabaj commented on packages/patternfly-4/react-core/src/components/Chip/Chip.js:
```diff
@@ -0,0 +1,57 @@
+  switch (variant) {
+    case ChipVariant.overflow:
+      return (
+        <ChipWrapper isOverflowed className={className} {...props}>
```
I don't know if the chip wrapper class is really needed.  Is there a reason we can't have the div here in the Chip Class?  Same Question goes for the ChipText.

boaz0 commented:
@tlabaj thanks. I removed `ChipText` and `ChipWrapper` as you suggested and used `getUniqueId` to set a default id.

@mcarrano yes. I will add `ChipGroup` later. I also did as you suggested: adding tooltip.

@mcoker thanks. I removed `aria-label`. Let me know if I'm missing anything.

Thank you all.

jschuler commented on packages/patternfly-4/react-core/src/components/Chip/Chip.d.ts:
```diff
@@ -0,0 +1,20 @@
+};
+
+export interface ChipProps extends Omit<HTMLProps<HTMLDivElement>, 'children' > {
+  children: string;
```
Can you get rid of the Omit 'children' and remove children from the interface? Also don't need id and onClick I believe since they are included in HTMLProps

jschuler commented on packages/patternfly-4/react-core/src/components/Chip/Chip.js:
```diff
@@ -0,0 +1,69 @@
+          </ChipButton>
+        </div>
+      );
+      return children.length < 16 ? (
```
Curious why 16? Should this be customizable?

ibolton336 commented:
@boaz0 I have opened a new PR for this as we need to expedite the process. I was unable to find a way to contact you but perhaps we can meet & pair program if your are in the Raleigh area this week. Thank you for your hard work on this. I have kept all of your commits intact & just cherry picked from your branch.

boaz0 commented:
@ibolton336 no problem.

Although, I am in Raleigh this week, I don't think I have the time to join in.

Thanks for the invitation, though. :grinning:
